libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,161 @@
+template <typename T>
+using remove_cvref_t = std::remove_cv_t<std::remove_reference_t<T>>;
+
+/// A wrapper around a member function pointer which (unlike std::mem_fn) allows
+/// calls to the function via a static function pointer allowing it to be
+/// called from Numba generated code.
+///
+/// @tparam Cls The class containing the method.
+/// @tparam Return The return value of the method.
+/// @tparam Args The argument types to the method.
+template <typename Cls, typename Return, typename... Args>
+struct MemberFunction {
+  Return (Cls::*member_func)(Args...);
+
+  static remove_cvref_t<Return> call(
+      MemberFunction<Cls, Return, Args...>* func, Cls* self,
+      remove_cvref_t<Args>... args) {
+    return (self->*(func->member_func))(args...);
+  }
+};
+
+template <typename Cls, typename Return, typename... Args>
+struct ConstMemberFunction {
+  Return (Cls::*member_func)(Args...) const;
+
+  static remove_cvref_t<Return> call(
+      ConstMemberFunction<Cls, Return, Args...>* func, const Cls* self,
+      remove_cvref_t<Args>... args) {
+    return (self->*(func->member_func))(args...);
+  }
+};
+
+template <typename... Args>
+class DefWithNumbaImpl {
+  /// Define an actual python method and numba wrapper using the provided
+  /// function pointer and data pointer (@p caller).
+  template <
+      typename Cls, typename Return, typename Func, typename Caller,
+      typename... Extra>
+  void def_class_method(
+      pybind11::class_<Cls>& cls, const char* name, Func f, Caller* caller,
+      const Extra&... extra) const {
+    cls.def(name, f, extra...);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_method")(
+        cls, cls.attr(name), (uintptr_t)&Caller::call, (uintptr_t)caller,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+  /// Define an actual python funct8ion and numba wrapper using the provided
+  /// function pointer. Unlike methods, there is no need for a data pointer.
+  template <typename Return, typename... Extra>
+  void def_func(
+      pybind11::module_& m, const char* name, Return (*f)(Args...),
+      const Extra&... extra) const {
+    m.def(name, f, extra...);
+    auto func = m.attr(name);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_function")(
+        func, (uintptr_t)f, 0,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+public:
+  // Overload to handle non-capturing lambdas and static functions. These
+  // should all take self as the first argument.
+  template <typename Return, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Return (*pf)(Args...),
```

Comment:
Are you assuming that `Extra...` and `Args...` match in types?

Reply:
No. `Extra...` is a bunch of "option" arguments that are used by pybind11. This code never interprets them. I just pass them through. The are only used to configure the generated python wrapper with things like doc strings and argument names and default argument values and stuff.